crates/networking/rpc/engine/fork_choice.rs

```diff
@@ -4,72 +4,152 @@ use ethrex_blockchain::{
 };
 
 #[derive(Debug)]
-pub struct ForkChoiceUpdatedV3 {
+pub enum ForkChoiceUpdatedVersion {
+    V1,
+    V2,
+    V3,
+}
+
+#[derive(Debug)]
+pub struct ForkChoiceUpdated {
     pub fork_choice_state: ForkChoiceState,
-    #[allow(unused)]
-    pub payload_attributes: Result<Option<PayloadAttributesV3>, String>,
+    pub payload_attributes: Result<Option<PayloadAttributes>, String>,
+    pub version: ForkChoiceUpdatedVersion,
 }
 
-impl TryFrom<ForkChoiceUpdatedV3> for RpcRequest {
-    type Error = String;
+impl ForkChoiceUpdated {
+    fn method(&self) -> String {
+        match self.version {
+            ForkChoiceUpdatedVersion::V1 => "engine_forkchoiceUpdatedV1".to_string(),
+            ForkChoiceUpdatedVersion::V2 => "engine_forkchoiceUpdatedV2".to_string(),
+            ForkChoiceUpdatedVersion::V3 => "engine_forkchoiceUpdatedV3".to_string(),
+        }
+    }
 
-    fn try_from(val: ForkChoiceUpdatedV3) -> Result<Self, Self::Error> {
-        match val.payload_attributes {
-            Ok(attrs) => Ok(RpcRequest {
-                method: "engine_forkchoiceUpdatedV3".to_string(),
-                params: Some(vec![
-                    serde_json::json!(val.fork_choice_state),
-                    serde_json::json!(attrs),
-                ]),
-                ..Default::default()
-            }),
-            Err(err) => Err(err),
+    fn version(&self) -> u8 {
+        match self.version {
+            ForkChoiceUpdatedVersion::V1 => 1u8,
+            ForkChoiceUpdatedVersion::V2 => 2u8,
+            ForkChoiceUpdatedVersion::V3 => 3u8,
         }
     }
-}
 
-impl RpcHandler for ForkChoiceUpdatedV3 {
-    // TODO(#853): Allow fork choice to be executed even if fork choice updated v3 was not correctly parsed.
-    fn parse(params: &Option<Vec<Value>>) -> Result<Self, RpcErr> {
+    fn validate(
+        &self,
+        attributes: &PayloadAttributes,
+        chain_config: ChainConfig,
+        head_block: BlockHeader,
+    ) -> Result<(), RpcErr> {
+        match self.version {
+            ForkChoiceUpdatedVersion::V1 => {
+                // TODO
+                Ok(())
+            }
+            ForkChoiceUpdatedVersion::V2 => {
+                if attributes.parent_beacon_block_root.is_some() {
+                    return Err(RpcErr::InvalidPayloadAttributes(
+                        "forkChoiceV2 with Beacon Root".to_string(),
+                    ));
+                }
+                if !chain_config.is_shanghai_activated(attributes.timestamp) {
+                    return Err(RpcErr::UnsuportedFork(
+                        "forkChoiceV2 used to build pre-Shanghai payload".to_string(),
+                    ));
+                }
+                if chain_config.is_cancun_activated(attributes.timestamp) {
+                    return Err(RpcErr::UnsuportedFork(
+                        "forkChoiceV2 used to build Cancun payload".to_string(),
+                    ));
+                }
```

Comment:
I think these can be avoided by replacing the optionals with the same "base struct wrapped with specific struct" technique. Then there are no extraneous attributes by construction, and the others can simply count on them.

Reply:
I think in this case I prefer to use `Optionals` for a couple of reasons:
- It's easier for serde to parse the request into the struct. The hive test checks if we receive a v2 request with v3 parameters and the other way around, so we need to allow v2 rpc with v3 params to signal that specific error. If we used the "base struct wrapped with specific struct" scheme, we would otherwise return a generic `RpcErr::BadParams` error (because we wouldn't be able to parse v3 params in a v2 call), but the hive test expects a `RpcErr::InvalidPayloadAttributes` error instead 🤦‍♂️
- The resulting structure (in this case a `BlockHeader`) also expects `Optionals` for the v3 params (ie. `parent_beacon_block_root`. So it is easier to set it directly in the end, as optional. See [here](https://github.com/lambdaclass/ethrex/blob/a969b8235c4bddf385ffd830654df6a8fd0f8621/crates/blockchain/payload.rs#L108)